File: SwiftBoy/cpp/Memory.cpp
//
//  Memory.cpp
//  SwiftBoy
//
//  Created by Andrew Gilliland on 12/21/25.
//

#include "Memory.hpp"
#include <iostream>
#include <cstring>

Memory::Memory() : romSize(0)
{
    std::cout << "[Memory] Constructor called" << std::endl;

    // Initialize all memory regions to 0
    rom.fill(0);
    vram.fill(0);
    wram.fill(0);
    oam.fill(0);
    hram.fill(0);
    io.fill(0);
}

bool Memory::loadROM(const uint8_t *data, std::size_t size)
{
    std::cout << "[Memory] Loading ROM: " << size << " bytes" << std::endl;

    if (data == nullptr || size == 0)
    {
        std::cerr << "[Memory] ERROR: Invalid ROM data" << std::endl;
        return false;
    }

    if (size > rom.size())
    {
        std::cerr << "[Memory] WARNING: ROM size (" << size
                  << ") exceeds maximum (" << rom.size()
                  << "), truncating..." << std::endl;
        size = rom.size();
    }

    // Copy ROM data into memory
    std::memcpy(rom.data(), data, size);
    romSize = size;

    // Print first few bytes for verification
    std::cout << "[Memory] First ROM bytes: "
              << std::hex
              << "0x" << (int)rom[0] << " "
              << "0x" << (int)rom[1] << " "
              << "0x" << (int)rom[2] << " "
              << "0x" << (int)rom[3]
              << std::dec << std::endl;

    std::cout << "[Memory] ROM loaded successfully" << std::endl;
    return true;
}

uint8_t Memory::read(uint16_t address)
{
    // ROM (0x0000-0x7FFF)
    if (address < 0x8000)
    {
        return rom[address];
    }
    // VRAM (0x8000-0x9FFF)
    else if (address < 0xA000)
    {
        return vram[address - 0x8000];
    }
    // External RAM (0xA000-0xBFFF) - Not implemented yet
    else if (address < 0xC000)
    {
        return 0xFF; // Return open bus value
    }
    // Work RAM (0xC000-0xDFFF)
    else if (address < 0xE000)
    {
        return wram[address - 0xC000];
    }
    // Echo RAM (0xE000-0xFDFF) - Mirrors 0xC000-0xDDFF
    else if (address < 0xFE00)
    {
        return wram[address - 0xE000];
    }
    // OAM (0xFE00-0xFE9F)
    else if (address < 0xFEA0)
    {
        return oam[address - 0xFE00];
    }
    // Unusable memory (0xFEA0-0xFEFF)
    else if (address < 0xFF00)
    {
        return 0xFF;
    }
    // I/O Registers (0xFF00-0xFF7F)
    else if (address < 0xFF80)
    {
        return io[address - 0xFF00];
    }
    // High RAM (0xFF80-0xFFFE)
    else if (address < 0xFFFF)
    {
        return hram[address - 0xFF80];
    }
    // Interrupt Enable Register (0xFFFF)
    else
    {
        return hram[0x7F]; // Store at end of HRAM
    }
}

void Memory::write(uint16_t address, uint8_t value)
{
    // ROM (0x0000-0x7FFF) - Read-only, but used for MBC control
    if (address < 0x8000)
    {
        // TODO: Implement MBC (Memory Bank Controller) support
        std::cout << "[Memory] ROM write ignored: 0x"
                  << std::hex << address << " = 0x" << (int)value
                  << std::dec << std::endl;
        return;
    }
    // VRAM (0x8000-0x9FFF)
    else if (address < 0xA000)
    {
        vram[address - 0x8000] = value;
    }
    // External RAM (0xA000-0xBFFF) - Not implemented yet
    else if (address < 0xC000)
    {
        // TODO: Implement cartridge RAM
    }
    // Work RAM (0xC000-0xDFFF)
    else if (address < 0xE000)
    {
        wram[address - 0xC000] = value;
    }
    // Echo RAM (0xE000-0xFDFF) - Mirrors 0xC000-0xDDFF
    else if (address < 0xFE00)
    {
        wram[address - 0xE000] = value;
    }
    // OAM (0xFE00-0xFE9F)
    else if (address < 0xFEA0)
    {
        oam[address - 0xFE00] = value;
    }
    // Unusable memory (0xFEA0-0xFEFF)
    else if (address < 0xFF00)
    {
        // Ignored
    }
    // I/O Registers (0xFF00-0xFF7F)
    else if (address < 0xFF80)
    {
        io[address - 0xFF00] = value;
    }
    // High RAM (0xFF80-0xFFFE)
    else if (address < 0xFFFF)
    {
        hram[address - 0xFF80] = value;
    }
    // Interrupt Enable Register (0xFFFF)
    else
    {
        hram[0x7F] = value;
    }
}

File: SwiftBoy/GameBoy/Memory/Memory.swift
//
//  Memory.swift
//  SwiftBoy
//
//  Created by Andrew Gilliland on 12/25/25.
//

import Foundation

class Memory {
    // Memory regions
    private var rom: [UInt8]      // 0x0000-0x7FFF (32KB ROM)
    private var vram: [UInt8]     // 0x8000-0x9FFF (8KB Video RAM)
    private var wram: [UInt8]     // 0xC000-0xDFFF (8KB Work RAM)
    private var oam: [UInt8]      // 0xFE00-0xFE9F (Sprite Attribute Table)
    private var io: [UInt8]       // 0xFF00-0xFF7F (I/O Registers)
    private var hram: [UInt8]     // 0xFF80-0xFFFE (High RAM)
    private var interruptEnable: UInt8  // 0xFFFF
    
    // ROM metadata
    private var romSize: Int = 0
    
    // Memory sizes
    private static let romSize = 0x8000    // 32KB
    private static let vramSize = 0x2000   // 8KB
    private static let wramSize = 0x2000   // 8KB
    private static let oamSize = 0x00A0    // 160 bytes
    private static let ioSize = 0x0080     // 128 bytes
    private static let hramSize = 0x007F   // 127 bytes
    
    init() {
        print("[Memory] Initializing...")
        
        // Initialize all memory regions
        self.rom = [UInt8](repeating: 0, count: Self.romSize)
        self.vram = [UInt8](repeating: 0, count: Self.vramSize)
        self.wram = [UInt8](repeating: 0, count: Self.wramSize)
        self.oam = [UInt8](repeating: 0, count: Self.oamSize)
        self.io = [UInt8](repeating: 0, count: Self.ioSize)
        self.hram = [UInt8](repeating: 0, count: Self.hramSize)
        self.interruptEnable = 0
        
        print("[Memory] Initialized")
    }
    
    deinit {
        print("[Memory] Deallocating")
    }
    
    // MARK: - ROM Loading
    
    func loadROM(_ data: Data) -> Bool {
        print("[Memory] Loading ROM: \(data.count) bytes")
        
        guard !data.isEmpty else {
            print("[Memory] ERROR: Empty ROM data")
            return false
        }
        
        // Check ROM size and truncate if necessary
        let bytesToCopy = min(data.count, Self.romSize)
        if data.count > Self.romSize {
            print("[Memory] WARNING: ROM size (\(data.count)) exceeds maximum (\(Self.romSize)), truncating...")
        }
        
        // Copy ROM data
        data.prefix(bytesToCopy).enumerated().forEach { index, byte in
            rom[index] = byte
        }
        romSize = bytesToCopy
        
        // Print first few ROM bytes for debugging
        let firstBytes = rom.prefix(4).map { String(format: "0x%02x", $0) }.joined(separator: " ")
        print("[Memory] First ROM bytes: \(firstBytes)")
        
        // Print ROM entry point
        let entryBytes = rom[0x0100..<0x0104].map { String(format: "0x%02x", $0) }.joined(separator: " ")
        print("[Memory] ROM entry point (0x0100): \(entryBytes)")
        
        print("[Memory] ROM loaded successfully")
        return true
    }
    
    // MARK: - Memory Access
    
    func read(address: UInt16) -> UInt8 {
        switch address {
        case 0x0000...0x7FFF:  // ROM
            return rom[Int(address)]
            
        case 0x8000...0x9FFF:  // VRAM
            return vram[Int(address - 0x8000)]
            
        case 0xA000...0xBFFF:  // External RAM (TODO: implement)
            return 0xFF
            
        case 0xC000...0xDFFF:  // WRAM
            return wram[Int(address - 0xC000)]
            
        case 0xE000...0xFDFF:  // Echo RAM (mirror of WRAM)
            return wram[Int(address - 0xE000)]
            
        case 0xFE00...0xFE9F:  // OAM
            return oam[Int(address - 0xFE00)]
            
        case 0xFEA0...0xFEFF:  // Unusable
            return 0xFF
            
        case 0xFF00...0xFF7F:  // I/O Registers
            return io[Int(address - 0xFF00)]
            
        case 0xFF80...0xFFFE:  // HRAM
            return hram[Int(address - 0xFF80)]
            
        case 0xFFFF:  // Interrupt Enable Register
            return interruptEnable
            
        default:
            print("[Memory] WARNING: Invalid read at 0x\(String(address, radix: 16))")
            return 0xFF
        }
    }
    
    func write(address: UInt16, value: UInt8) {
        switch address {
        case 0x0000...0x7FFF:  // ROM (read-only, but used for MBC control)
            // TODO: Handle MBC register writes
            break
            
        case 0x8000...0x9FFF:  // VRAM
            vram[Int(address - 0x8000)] = value
            
        case 0xA000...0xBFFF:  // External RAM (TODO: implement)
            break
            
        case 0xC000...0xDFFF:  // WRAM
            wram[Int(address - 0xC000)] = value
            
        case 0xE000...0xFDFF:  // Echo RAM (mirror of WRAM)
            wram[Int(address - 0xE000)] = value
            
        case 0xFE00...0xFE9F:  // OAM
            oam[Int(address - 0xFE00)] = value
            
        case 0xFEA0...0xFEFF:  // Unusable
            break
            
        case 0xFF00...0xFF7F:  // I/O Registers
            io[Int(address - 0xFF00)] = value
            
        case 0xFF80...0xFFFE:  // HRAM
            hram[Int(address - 0xFF80)] = value
            
        case 0xFFFF:  // Interrupt Enable Register
            interruptEnable = value
            
        default:
            print("[Memory] WARNING: Invalid write at 0x\(String(address, radix: 16))")
        }
    }
    
    // MARK: - Direct Access
    
    func getVRAM() -> UnsafePointer<UInt8> {
        return vram.withUnsafeBufferPointer { $0.baseAddress! }
    }
}
